src/functionalTest/java/uk/gov/hmcts/reform/pbis/integration/MessageQueueProcessorTest.java

```diff
@@ -0,0 +1,193 @@
+        IServiceBusClientFactory clientFactorySpy = prepareServiceBusClientFactorySpy();
+        messageQueueProcessor = new MessageQueueProcessor(clientFactorySpy, emailService);
+    }
+
+    @Override
+    @After
+    public void tearDown() throws Exception {
+        super.tearDown();
+    }
+
+    @Test
+    public void run_should_read_all_messages_from_subscription() throws Exception {
+        List<PrivateBetaRegistration> registrations = Stream
+            .generate(() -> SampleData.getSampleRegistration(testConfig.getServiceName()))
+            .limit(5)
+            .collect(Collectors.toList());
+
+        for (PrivateBetaRegistration registration : registrations) {
+            serviceBusFeeder.sendMessage(registration);
+        }
+
+        messageQueueProcessor.run();
+
+        // check if there's nothing left to read from the subscription
+        assertThat(receiveMessage()).isNull();
+    }
+
+    @Test
+    public void run_should_call_email_service_for_each_valid_message() throws Exception {
+        int validMesssageCount = 3;
+
+        List<PrivateBetaRegistration> registrationsToProcess =
+            SampleData.getSampleRegistrations(testConfig.getServiceName(), validMesssageCount);
+
+        serviceBusFeeder.sendMessage(registrationsToProcess.get(0));
+        serviceBusFeeder.sendMessage("invalid1");
+        serviceBusFeeder.sendMessage(registrationsToProcess.get(1));
+        serviceBusFeeder.sendMessage("invalid2");
+        serviceBusFeeder.sendMessage(registrationsToProcess.get(2));
+
+        messageQueueProcessor.run();
+
+        ArgumentCaptor<PrivateBetaRegistration> registrationCaptor =
+            ArgumentCaptor.forClass(PrivateBetaRegistration.class);
+
+        verify(emailService, times(validMesssageCount))
+            .sendWelcomeEmail(registrationCaptor.capture());
+
+        String[] actualReferenceIds = getRegistrationIds(registrationCaptor.getAllValues());
+        String[] expectedReferenceIds = getRegistrationIds(registrationsToProcess);
+
+        assertThat(actualReferenceIds).containsExactlyInAnyOrder(expectedReferenceIds);
+    }
+
+    @Test
+    public void run_should_not_call_email_service_for_invalid_messages() throws Exception {
+        serviceBusFeeder.sendMessages("invalid1", "invalid2", "invalid3");
+
+        messageQueueProcessor.run();
+
+        verify(emailService, never()).sendWelcomeEmail(any());
+    }
+
+    @Test
+    public void run_should_consume_valid_message_so_they_are_not_reprocessed() throws Exception {
+        final String invalidMessageContent = "Invalid content";
+
+        serviceBusFeeder.sendMessage(
+            SampleData.getSampleRegistration(testConfig.getServiceName())
+        );
+
+        serviceBusFeeder.sendMessage(invalidMessageContent);
+
+        messageQueueProcessor.run();
+
+        assertThat(receiveMessage()).isNull();
+
+        // wait for message lock to expire
+        Thread.sleep(testConfig.getMessageLockTimeoutInMs());
```

Comment:
Integration tests will have their own subscription (ideally dynamically created). This subscription will have a very short message lock timeout.